gg_fit_curve.py

```diff
@@ -0,0 +1,402 @@
+"""
+Python transcription of https://github.com/erich666/GraphicsGems/blob/master/gems/FitCurves.c
+as it existed 2/24/2020.
+"""
+from math import sqrt
+from typing import NamedTuple, Sequence, Tuple
+
+
+class Point(NamedTuple):
+    x: float
+    y: float
+
+    def magnitude(self) -> float:
+        return sqrt(self.x * self.x + self.y * self.y)
+
+    # V2AddII; it's weird this works on vectors...
+    def add(self, other) -> "Point":
+        return Point(self.x + other.x, self.y + other.y)
+
+    # V2SubII
+    def sub(self, other) -> "Point":
+        return Point(self.x - other.x, self.y - other.y)
+
+    def scale(self, s: float) -> "Point":
+        return Point(self.x * s, self.y * s)
+
+    @staticmethod
+    def tan_left(points: Sequence["Point"]) -> "Vector":
+        assert len(points) > 1
+        return Vector.p0_to_p1(points[0], points[1]).normalize()
+
+    @staticmethod
+    def tan_right(points: Sequence["Point"]) -> "Vector":
+        assert len(points) > 1
+        return Vector.p0_to_p1(points[-1], points[-2]).normalize()
+
+    @staticmethod
+    def tan_center(points: Sequence["Point"]) -> "Vector":
+        assert len(points) == 3
+        v1 = Vector.p0_to_p1(points[1], points[0])
+        v2 = Vector.p0_to_p1(points[2], points[1])
+        tHatCenter = Vector((v1.x + v2.x) / 2, (v1.y + v2.y) / 2)
+        return tHatCenter.normalize()
+
+
+CubicBezier = Tuple[Point, Point, Point, Point]
+
+
+class Vector(Point):
+    def squared_length(self) -> float:
+        return self.x * self.x + self.y * self.y
+
+    def length(self) -> float:
+        return sqrt(self.squared_length())
+
+    def normalize(self) -> "Vector":
+        length = self.length()
+        x, y = 0.0, 0.0
+        if length != 0.0:
+            x = self.x / length
+            y = self.y / length
+
+        return Vector(x, y)
+
+    def negate(self) -> "Vector":
+        return Vector(-self.x, -self.y)
+
+    @staticmethod
+    def p0_to_p1(p0: Point, p1: Point) -> "Vector":
+        return Vector(*(p1.sub(p0)))
```

Comment:
subtracting two points to obtain a vector `vec = (end_pt - start_pt)` looks clearer to me, maybe I just got used to reading that way.

Reply:
Fixing while migrating to picosvg